src/psyclone/psyad/adjoint_visitor.py

```diff
@@ -132,6 +133,32 @@ def schedule_node(self, node):
         node_copy = node.copy()
         node_copy.children = []
 
+        if isinstance(node.parent, (Routine, Container)):
```

Comment:
If I've understood this correctly, this could add an Assignment node directly to a Container which is incorrect (but would probably be caught by the parent-child validity checks)?

Reply:
Oops, that logic was wrong. I think the test should be whether the node itself is a Routine.